File: Assets/scripts/Player.cs
using UnityEngine;
using System.Collections;

public class Player : MonoBehaviour {
	CharacterController cc;
	CollisionFlags cf;
	
	public float speed = 10f;
	public float jumpSpeed = 9f;
	public float gravity = 21f;
	public float terminalVelocity = 20f;
	
	public Vector3 moveVector;
	public float verticalVelocity;

	// Use this for initialization
	void Start() {
		cc = GetComponent<CharacterController>();
	}
	
	// Update is called once per frame
	void Update() {
		checkMovement();
		handleActionInput();
		processMovement();
	}
	
	void checkMovement(){
		//	Move left/right.
		var deadZone = 0.1f;
		verticalVelocity = moveVector.y;
		
		if(isCeiled()){
			verticalVelocity = 0;
		}
		
		moveVector = Vector3.zero;
		
		if(Input.GetAxis("Horizontal") > deadZone || Input.GetAxis("Horizontal") < -deadZone){
			moveVector += new Vector3(Input.GetAxis("Horizontal"), 0, 0);
		}
	}
	
	void handleActionInput(){
		if(Input.GetButton("Jump")){
			jump();
		}
	}
	
	void processMovement(){
		//	Transform moveVector into world-space relative to character rotation.
		moveVector = transform.TransformDirection(moveVector);
		
		//	Normalized moveVector if magniture > 1.
		if(moveVector.magnitude > 1){
			moveVector = Vector3.Normalize(moveVector);
		}
		
		moveVector *= speed;
		
		//	Reapply vertical vlocity to moveVector.y.
		moveVector = new Vector3(moveVector.x, verticalVelocity, moveVector.z);
		
		applyGravity();
		
		cf = cc.Move(moveVector * Time.deltaTime);
	}
	
	void applyGravity(){
		if(moveVector.y > -terminalVelocity){
			moveVector = new Vector3(moveVector.x, (moveVector.y - gravity * Time.deltaTime), moveVector.z);
		}
		if(cc.isGrounded && moveVector.y < -1){
			moveVector = new Vector3(moveVector.x, -1, moveVector.z);
		}
	}
	
	public void jump(){
		if(cc.isGrounded){
			verticalVelocity = jumpSpeed;
		}
	}
	
	bool isGrounded() {
		return cc.isGrounded;
		//return (cf & CollisionFlags.CollidedBelow) != 0;
	}
	
	bool isCeiled() {
		return (cf & CollisionFlags.CollidedAbove) != 0;	
	}
	
	bool isVerticallyCollided() {
		return isGrounded() || isCeiled();
	}
}

File: Assets/scripts/Player.js
#pragma strict
 
////////////////////////////////
//	 Notes
////////////////////////////////
 
/*
 *	 This is a boilerplate controller for a 2D or 2.5D platform game.
 *	 To that end, this controller allows for movement along the X and Y
 *	 axes only. There is built-in support for jumping, double jumping,
 *	 rotating in 3D (as opposed to 2D), setting a spawn point, etc.
 *	 This script was influenced by Unity's 2D Gameplay Tutorial:
 *	 http://unity3d.com/support/resources/tutorials/2d-gameplay-tutorial.
 */
  
////////////////////////////////////////////////////////////////
//	 Requires
////////////////////////////////
 
// Require a character controller to be attached to the same game object
@script RequireComponent(CharacterController)
 
////////////////////////////////
//	 Variables
////////////////////////////////
 
var spawnPoint : Transform; // The character will spawn here
var movement : Movement;
var collision : CollisionFlags;
var state : int;
 
private var character : CharacterController;

////////////////////////////////
//	 Classes
////////////////////////////////
 
class Movement {
	////////////////////////////////
	//	 Public (Inspector) variables
	////////////////////////////////
 
	var enabled : boolean = true; // Is the character controller enabled?
	var gravity : float = 20;
	var jumpHeight : float = 14;
	var walkSpeed : float = 6;
	var runSpeed : float = 10;
	var rotate : boolean = true; // Should the character rotate?
	var rotateIn3D : boolean = true; // Should the character be rotated in 3D?
	var rotationSmoothing : float = 10; // Rotation Smoothing speed (for "Rotate In 3D")
 
	////////////////////////////////
	//	 NonSerialized variables
	////////////////////////////////
 
	// The character's current horizontal direction
	@System.NonSerialized
	var horizontalDirection : int;
 
	// The character's current direction (on all planes)
	@System.NonSerialized
	var direction : Vector3;
 
	// The character's current movement offset
	@System.NonSerialized
	var offset : Vector3;
 
	////////////////////////////////
	//	 Jumping
	////////////////////////////////
 
	// Did the user press the jump input button while midair?
	@System.NonSerialized
	var usedExtraJump : boolean = false;
}
 
////////////////////////////////
//	 Unity Functions
////////////////////////////////
 
function Awake() {
	character = GetComponent(CharacterController);
	state = 0;
	Spawn();
}
 
function FixedUpdate() {
	// Make sure the character stays in the 2D plane
	transform.position.z = 0;
}
 
function Update() {
	// GetAxisRaw("Horizontal") will return either -1 or 1 while the left or right input is depressed
	// Otherwise, GetAxisRaw("Horizontal") returns 0 for no input
	movement.horizontalDirection = (movement.enabled) ? Input.GetAxisRaw("Horizontal") : 0;
	if (movement.horizontalDirection != 0) movement.direction = Vector3(movement.horizontalDirection, 0, 0);
	MoveCharacter();
}
 
////////////////////////////////
//	 Custom Functions
////////////////////////////////
 
function Spawn() {
	transform.position = spawnPoint.position;
}
 
// Move the character using Unity's CharacterController.Move function
// http://docs.unity3d.com/Documentation/ScriptReference/CharacterController.Move.html
function MoveCharacter() {
	ApplyMovement();
	ApplyGravity();
	// Move the character (with deltaTime to ensure frame rate independence)
	collision = character.Move(movement.offset * Time.deltaTime);
}
 
function ApplyMovement() {
	var speed = movement.walkSpeed;
	
	switch (character.isGrounded) {
		// The character is on the ground
		case true:
			movement.usedExtraJump = false;
			if (movement.enabled) {
				var isRunning = Input.GetButton("Run");
				if(isRunning){
					speed = movement.runSpeed;
				}
				
				movement.offset = Vector3(Input.GetAxis("Horizontal"), 0, 0) * speed;
				
				if(Input.GetButtonDown("Jump")){
					ApplyJump();
				}
			} else {
				movement.offset = Vector3.zero;
			}
		break;
 
		// The character is midair
		case false:
			if (movement.enabled) {
				movement.offset.x = Input.GetAxis("Horizontal") * speed;
				// Apply an extra jump if the jump input button is pressed a second time
				// The final jump height will be greatest at the apex of the first jump
				if (Input.GetButtonDown("Jump") && ! movement.usedExtraJump) {
					movement.usedExtraJump = true;
					// ApplyJump();
				}
				
				if(isCeiled()){
					movement.offset.y = -speed;
				}
			}
		break;
	}
 
	if (movement.rotate) ApplyRotation();
}
 
function ApplyJump() {
	movement.offset.y = movement.jumpHeight;
}
 
function ApplyRotation() {
	if (movement.direction.x != 0) {
		if (movement.rotateIn3D) {
			// http://docs.unity3d.com/Documentation/ScriptReference/Transform-rotation.html
			transform.rotation = Quaternion.Slerp(transform.rotation, Quaternion.LookRotation(movement.direction), Time.deltaTime * movement.rotationSmoothing);
		} else {
			// http://docs.unity3d.com/Documentation/ScriptReference/Quaternion.LookRotation.html
			transform.rotation = Quaternion.LookRotation(movement.direction);
		}
	}
}
 
function ApplyGravity() {
	movement.offset.y -= movement.gravity * Time.deltaTime;
}
	
function isGrounded() : boolean {
	return character.isGrounded;
}

function isCeiled() : boolean {
	return (collision & CollisionFlags.CollidedAbove) != 0;	
}

function grow() {
	state = 1;
}
